classes/plugininfo/factor.php

```diff
@@ -82,6 +82,48 @@ public static function get_enabled_factors() {
         return $return;
     }
 
+    /**
+     * Finds enabled factors for user by userid.
+     *
+     * @param int $userid user id.
+     * @return array of enabled factors for given user.
+     */
+    public static function get_enabled_user_factors($userid) {
+        $return = array();
+        $factors = self::get_enabled_factors();
+
+        foreach ($factors as $factor) {
+            $userfactors = $factor->get_enabled_user_factor($userid);
+            if ($userfactors) {
+                $return[$factor->name] = $userfactors;
+            }
+        }
+
+        return $return;
+    }
+
+    /**
+     * Gets next factor to authenticate user.
+     *
+     * @param int $userid user id.
+     * @return string name of the next factor to be authenticated.
+     */
+    public static function get_next_user_factor($userid) {
+        $factors = self::get_enabled_factors();
+
+        foreach ($factors as $factor) {
+            $userfactor = $factor->get_enabled_user_factor($userid);
+            if ($userfactor) {
+                $property = 'factor_'.$userfactor->name.'_authenticated';
+                if (empty($_SESSION['USER']->$property) || true !== $_SESSION['USER']->$property) {
+                    return $userfactor->name;
```

Comment:
why not just pass back the factor object instead?

Reply:
Corrected